crates/rolldown/src/utils/renamer.rs

```diff
@@ -34,65 +27,149 @@ pub struct Renamer<'name> {
-      OutputFormat::Cjs => vec!["module", "require", "__filename", "__dirname", "exports"],
-      OutputFormat::Iife | OutputFormat::Umd => vec!["exports"], // Also for  AMD, but we don't support them yet.
+      OutputFormat::Esm | OutputFormat::App => FxHashSet::default(),
+      OutputFormat::Cjs => {
+        FxHashSet::from_iter(["module", "require", "__filename", "__dirname", "exports"])
+      }
+      OutputFormat::Iife | OutputFormat::Umd => FxHashSet::from_iter(["exports"]), // Also for  AMD, but we don't support them yet.
     };
     // https://github.com/rollup/rollup/blob/bfbea66569491f5466fbba99de2ba6a0225f851b/src/Chunk.ts#L1359
     manual_reserved.extend(["Object", "Promise"]);
+
     Self {
-      canonical_names: FxHashMap::default(),
-      canonical_token_to_name: FxHashMap::default(),
-      symbol_db: symbols,
       used_canonical_names: manual_reserved
         .iter()
         .chain(RESERVED_KEYWORDS.iter())
         .chain(GLOBAL_OBJECTS.iter())
         .map(|s| (Rstr::new(s), 0))
         .collect(),
+      manual_reserved,
+      canonical_names: FxHashMap::default(),
+      canonical_token_to_name: FxHashMap::default(),
+      symbol_db,
+      module_used_names: base_module_index
+        .map(|index| FxHashMap::from_iter([(index, FxHashSet::default())]))
+        .unwrap_or_default(),
+      entry_module_used_names: base_module_index
+        .map(|index| {
+          symbol_db.local_db(index).ast_scopes.scoping().symbol_names().collect::<FxHashSet<_>>()
+        })
+        .unwrap_or_default(),
+      entry_module: base_module_index.map(|index| symbol_db.local_db(index)),
+      used_names: FxHashSet::default(),
     }
   }
 
   pub fn reserve(&mut self, name: Rstr) {
     self.used_canonical_names.insert(name, 0);
   }
 
+  #[expect(clippy::map_entry)]
   pub fn add_symbol_in_root_scope(&mut self, symbol_ref: SymbolRef) {
     let canonical_ref = symbol_ref.canonical_ref(self.symbol_db);
     let original_name = canonical_ref.name(self.symbol_db).to_rstr();
-    match self.canonical_names.entry(canonical_ref) {
-      Entry::Vacant(vacant) => {
-        let mut candidate_name = original_name.clone();
-        loop {
-          match self.used_canonical_names.entry(candidate_name.clone()) {
-            Entry::Occupied(mut occ) => {
-              let next_conflict_index = *occ.get() + 1;
-              *occ.get_mut() = next_conflict_index;
-              candidate_name =
-                concat_string!(original_name, "$", itoa::Buffer::new().format(next_conflict_index))
-                  .into();
-            }
-            Entry::Vacant(vac) => {
-              vac.insert(0);
-              break;
-            }
-          }
-        }
-        vacant.insert(candidate_name);
+
+    if !self.canonical_names.contains_key(&canonical_ref) {
+      let name = self.get_unique_name(symbol_ref, &original_name);
+      self.canonical_names.insert(canonical_ref, name);
```

Comment:
Here the `insert` `canonical_names` will hash twice.

Reply:
I've changed this.